Feature/experienced wait time delay

WuestFlorian opened the pull request:
Closes https://github.com/SWT-Forschungsprojekt/planning/issues/24

This implements the `Experienced Wait Time Delay` metric specified [here](https://docs.google.com/document/d/1-AOtPaEViMcY6B5uTAYj7oVkwry3LfAQJg3ihSRTVoU).

Additionally, the estimation of actual arrival times and delays now only considers stop time updates with uncertainty 0. 
This is adapted in the accuracy metrics to not include updates without valid actual arrival time and delay.

maexled committed:
implement experienced wait time delay metric

TODO: check if the logic works correctly

WuestFlorian committed:
Change computation

Change computation of experienced wait time delay

WuestFlorian committed:
Fix multiple updates at last timestamps

Fixed that if multiple updates for a route and stop combination are published at the same time (e.g. for multiple trips), the one with the lowest arrival time is taken

WuestFlorian committed:
Adapt accuracy

Adapt accuracy to ignore stop time update where no actual arrival time i known

WuestFlorian committed:
Fix unknown actual arrival time/delay handling

Fixed handling of samples with unknown actual arrival time / delay, as previously also samples with value 0 were discarded

WuestFlorian committed:
Resolve merge conflicts

1Maxnet1 commented on src/realtime_metrics/main.py:
```diff
@@ -175,25 +185,151 @@ def eta_accuracy(stop_time_updates: list[tuple[TripUpdate, StopTimeUpdate]]):
+    Computes the average Experienced Wait Time Delay metrics for the given stop time updates.
+    The metric is defined here: https://docs.google.com/document/d/1-AOtPaEViMcY6B5uTAYj7oVkwry3LfAQJg3ihSRTVoU. 
+    It computes the average amount of time in minutes a passenger has to wait at a stop, 
+    if he arrives at the arrival time of the most up to date stop time update.
```
```suggestion
    if he arrives at the arrival time of the most up-to-date stop time update.
```

1Maxnet1 commented on src/realtime_metrics/main.py:
```diff
@@ -175,25 +185,151 @@ def eta_accuracy(stop_time_updates: list[tuple[TripUpdate, StopTimeUpdate]]):
+            delays.append(delay)
+
+            # Move to the next minute
+            current_time += 60
```
For me this really looks like a for-loop made out of a while-loop, so I would prefer an actual for loop
```suggestion
        for current_time in range(int(min_arrival_time), int(end_time) + 1, 60):
            # find the last stop time update
            next_predicted_arrival = get_last_predicted_update(current_time, trip_stop_updates)
            if not next_predicted_arrival:
                continue  # No prediction available
            next_predicted_arrival_time = next_predicted_arrival[1].arrival_time
            if next_predicted_arrival_time > current_time + 3600:
                continue  # prediction is more than 60 minutes in the future

            # find next experienced arrival at or after predicted 
            next_actual_arrival = get_next_actual_arrival(next_predicted_arrival_time, route_id, stop_id)
            if not next_actual_arrival:
                continue  # no actual arrival after the given timestamp found
            next_actual_arrival_time = next_actual_arrival.arrival_time

            delay = next_actual_arrival_time - next_predicted_arrival_time
            delays.append(delay)
```

1Maxnet1 commented on src/realtime_metrics/main.py:
```diff
@@ -175,25 +185,151 @@ def eta_accuracy(stop_time_updates: list[tuple[TripUpdate, StopTimeUpdate]]):
+        return None
+
+
+def experienced_wait_time_delay(trip_stop_time_updates: list[tuple[TripUpdate, StopTimeUpdate]]) -> float:
```
Could also return None
```suggestion
def experienced_wait_time_delay(trip_stop_time_updates: list[tuple[TripUpdate, StopTimeUpdate]]) -> floating[
                                                                                                        Any] | None:
```

1Maxnet1 reviewed (commented):
Looks good to me :+1: Left some minor comments on how to make the code more understandable

WuestFlorian committed:
Review suggestions

1Maxnet1 reviewed (changes requested):
When I run the script on the same database to calculate the metrics once with this branch and once with development, it will calculate 2 different Availability of acceptable stop time updates. As this PR is about EWTD, I would assume this should not change, no?

Also I could not create any result, where the EWTD could be calculated, as it always claim to have missing data. Did it work for you and if so with which feed, dataset and collection time combination?

WuestFlorian commented:
> When I run the script on the same database to calculate the metrics once with this branch and once with development, it will calculate 2 different Availability of acceptable stop time updates. As this PR is about EWTD, I would assume this should not change, no?

I tested it on both the main branch as well as on the EWTD branch, and in both cases got 0.0 as `availability of acceptable stop time updates`.  This is the correct value, as updates are published every minute (and roughly 20 seconds off, so there should be no minute slots with multiple updates). What values did you get?

> Also I could not create any result, where the EWTD could be calculated, as it always claim to have missing data. Did it work for you and if so with which feed, dataset and collection time combination?

For me, EWTD worked. I used the bucharest feed with 5 minutes worth of updates (from 13th march 2025, 15:38-15:42).

What feed and data range did you use for your testing?

1Maxnet1 commented:
> I tested it on both the main branch as well as on the EWTD branch, and in both cases got 0.0 as availability of acceptable stop time updates. This is the correct value, as updates are published every minute (and roughly 20 seconds off, so there should be no minute slots with multiple updates). What values did you get?

On the Bucharest feed I get the same values, so looks good to me. I might have tested another feed, but I would suggest we first merge the current state and when all metrics are implemented, we test with the different feeds to identify further issues. When I stumble upon an issue while developing, I can also report it as GitHub issue with how I reproduced it.

> For me, EWTD worked. I used the bucharest feed with 5 minutes worth of updates (from 13th march 2025, 15:38-15:42).

I tried again with the Bucharest feed for 5 minutes but still it claims no data provided.

> What feed and data range did you use for your testing?

I used a different one before, but now Bucharest with 5 Minutes from 10:20 to 10:25 today.

WuestFlorian committed:
Fix timestamp casting

Fixed casting datetime to unix timestamp to use utc instead of local time zone

WuestFlorian committed:
Fix stop time update handling order

First adding stop time updates to sorted route, trip and stop set and then update actual arrival times, to not skip it if arrival uncertainty is greater than zero

WuestFlorian commented:
I identified a problem with converting datetime to unix timestamps due to timezone mismatches, which caused most samples to be outside the valid range. Now, I get valid results with both time frames.

1Maxnet1 commented on src/realtime_metrics/main.py:
```diff
@@ -84,21 +95,26 @@ def mse_accuracy(stop_time_updates: list[tuple[TripUpdate, StopTimeUpdate]]):
 
     # compute MSE
-    sum = 0
+    if len(samples) <= 0:
```
length smaller then 0 is not possible, so I was confused when initially reading this. We might it replace it like this:
```suggestion
    if len(samples) == 0:
```

1Maxnet1 commented on src/realtime_metrics/main.py:
```diff
@@ -17,59 +18,69 @@ def run_analysis():
+    if experienced_wait_time_delay_result is None:
+        print("Experienced Wait Time Delay could not be computed, no data provided!")
+    else:
+        print("Experienced Wait Time Delay: ", experienced_wait_time_delay_result)
```
For the accuracy it is not needed, but Experienced wait time delay should give a unit, to know whether its in minutes or seconds or what

1Maxnet1 commented on src/realtime_metrics/main.py:
```diff
@@ -175,73 +195,204 @@ def eta_accuracy(stop_time_updates: list[tuple[TripUpdate, StopTimeUpdate]]):
+            if next_predicted_arrival_time > current_time + 3600:
+                continue  # prediction is more than 60 minutes in the future
+            # find next experienced arrival at or after predicted 
+            next_actual_arrival = get_next_actual_arrival(next_predicted_arrival_time, route_id, stop_id)
```
You claim that you expect ints, but you pass strings. As it seems to work, you might update the type hint in get_next_actual_arrival

1Maxnet1 commented on src/realtime_metrics/main.py:
```diff
@@ -175,73 +195,204 @@ def eta_accuracy(stop_time_updates: list[tuple[TripUpdate, StopTimeUpdate]]):
-    computes the availability of acceptable stop time updates metrics for the given stop time updates in the given time frame.
+    Computes the availability of acceptable stop time updates metrics for the given stop time updates in the given time frame.
+    The metric is defined here: https://docs.google.com/document/d/1-AOtPaEViMcY6B5uTAYj7oVkwry3LfAQJg3ihSRTVoU. 
+    It computes the percentage of one minute slots with two or more updates.
```
```suggestion
    It computes the percentage of one-minute slots with two or more updates.
```

1Maxnet1 reviewed (approved):
Works for me now too, thanks! I left some cosmetic comments. Feel free to include them but can be already be merged as is.

WuestFlorian committed:
Apply code review suggestions

WuestFlorian committed:
Merge branch 'main' into feature/experienced-wait-time-delay
